ofp4/tests/test.rs

```diff
@@ -0,0 +1,444 @@
+        let program = self.get_program().to_string_lossy().into_owned();
+        let mut cleanup = Cleanup::new()?;
+        let mut child = self.start(&mut cleanup)?;
+        let status = wait_for_child_to_die(&mut child)??;
+        cleanup.cancel();
+
+        info!("{program} exited ({})", status);
+        if !status.success() {
+            Err(anyhow!("{program} failed ({status})"))?;
+        }
+        Ok(())
+    }
+}
+
+fn trace_flow<'a, P, I, S>(tmp_dir: P, args: I) -> Result<(String, String)>
+where P: AsRef<Path>,
+      I: IntoIterator<Item = S>,
+      S: AsRef<OsStr>
+{
+    // Run the trace.
+    let mut command = ovs_command("ovs-appctl", &tmp_dir);
+    command.arg("ofproto/trace").arg("br0").args(args);
+    let command_string = command.command_string();
+    info!("Running {command_string}...");
+    let output = String::from_utf8(Cleanup::output(&mut command)?.stdout)?;
+
+    // Extract just the datapath actions from the last line and return them.
+    debug!("{output}");
+    let last_line = String::from(output.lines().nth_back(0).unwrap_or(""));
+    info!("...{last_line}");
+    if let Some(rest) = last_line.strip_prefix("Datapath actions: ") {
+        Ok((output, rest.into()))
+    } else {
+        Err(anyhow!("Trace command returned unexpected output:\n{command_string}\n{output}"))
+    }
+}
+
+const DEVICE_ID: u64 = 1;
+
+fn election_id() -> Uint128 {
+    Uint128 { high: 0, low: 1, ..Default::default() }
+}
+
+async fn start_ofp4(p4info: P4Info) -> Result<(Cleanup, PathBuf, P4RuntimeClient)> {
+    grpcio::redirect_log();
+    
+    let mut cleanup = Cleanup::new()?;
+    if let Ok(_) = std::env::var("KEEP_TMPDIR") {
+        cleanup.keep_temp_dirs();
+    }
+    let tmp_dir = cleanup.create_temp_dir(".")?;
+
+    // Create OVS configuration database.
+    ovs_command("ovsdb-tool", &tmp_dir).arg("create").arg("ovsdb.conf.db").run()?;
+
+    // Start ovsdb-server to serve the configuration database.
+    cleanup.register_pidfile(tmp_dir.join("ovsdb-server.pid"))?;
+    ovs_command("ovsdb-server", &tmp_dir).arg("ovsdb.conf.db").arg("--remote=punix:db.sock").arg("--pidfile").arg("--detach").run()?;
+
+    // Use ovs-vsctl to configure OVS.
+    let mut command = ovs_command("ovs-vsctl", &tmp_dir);
+    command.args(["--no-wait", "--", "add-br", "br0"]);
+    for port in 1..=4 {
+        let portname = format!("p{port}");
+        command.args(["--", "add-port", "br0", &portname,
+                      "--", "set", "Interface", &portname, &format!("ofport_request={port}")]);
+    }
+    command.run()?;
+
+    // Start ovs-vswitchd.
+    cleanup.register_pidfile(tmp_dir.join("ovs-vswitchd.pid"))?;
+    ovs_command("ovs-vswitchd", &tmp_dir)
+        .arg("--log-file").arg("-vvconn").arg("-vconsole:off")
+        .arg("--enable-dummy=override").arg("--disable-system").arg("--disable-system-route")
+        .arg("--detach").arg("--pidfile")
+        .arg("unix:db.sock")
+        .run()?;
```

Comment:
so if this fails, the previously started things are not cleaned-up?

Reply:
This is the magic of the `Cleanup` struct. If this fails and returns Err, then Cleanup already has pidfiles registered for long-running processes. When it gets dropped as part of returning Err, it will kill those processes and check that they really die, and it will delete all the files in the temporary directory and then the temporary directory itself. (If the function succeeds, this function returns the Cleanup to the caller and the caller can keep using it.)

Reply:
that's nice.

Reply:
The `Cleanup` killed and deletes everything for us automatically.